libkbfs/tlf_journal_test.go

```diff
@@ -791,3 +738,37 @@ func TestTLFJournalPreservesBranchID(t *testing.T) {
 	config.checkRange(rmdses, firstRevision, firstPrevRoot, Unmerged,
 		rmdses[0].MD.BID())
 }
+
+// TestTLFJournalFlushOrdering tests that we don't naively push all
+// block ops, and then push all MD ops, skipping any block ops that
+// were added in the meantime.
+func TestTLFJournalFlushOrdering(t *testing.T) {
+	tempdir, config, ctx, cancel, tlfJournal, delegate :=
+		setupTLFJournalTest(t, TLFJournalBackgroundWorkPaused)
+	defer teardownTLFJournalTest(
+		tempdir, config, ctx, cancel, tlfJournal, delegate)
+
+	firstRevision := MetadataRevision(10)
+	firstPrevRoot := fakeMdID(1)
+	mdCount := 10
+
+	prevRoot := firstPrevRoot
+	for i := 0; i < mdCount; i++ {
+		revision := firstRevision + MetadataRevision(i)
+		md := config.makeMD(revision, prevRoot)
+		mdID, err := tlfJournal.putMD(ctx, md)
+		require.NoError(t, err)
+		prevRoot = mdID
+	}
+
+	var mdserver shimMDServer
+	config.mdserver = &mdserver
+	mdserver.onPut = func() error {
+		putBlock(ctx, t, config, tlfJournal, []byte{1, 2, 3, 4})
+		return nil
+	}
+
+	err := tlfJournal.flush(ctx)
+	require.NoError(t, err)
+	requireJournalEntryCounts(t, tlfJournal, 0, 0)
```

Comment:
I think you might be overlooking the fact that now `flush` has *two* nested loops, whereas it previously just made one pass through. Now flush looks like:

```
while snapshot := getSnapshot(..); !snapshot.empty() {
  drain blockJournal(snapshot)
  drain mdJournal(snapshot)
}
```

whereas before it was just:

```
drain blockJournal
drain mdJournal
```

Does that help?